### PR TITLE
feat(react-scheduler): add appointment resources support

### PR DESCRIPTION
**Pull request by MaximKudriavtsev**
Appointments & AppointmentTooltip

image

AppointmentForm

image

### CODE REVIEW

File: packages/dx-scheduler-core/src/types/resources.types.ts [hunk 1]
```diff
@@ -0,0 +1,72 @@
+import { PureComputed } from '@devexpress/dx-core';
+import { Appointment } from './../index';
+
+export type PlaneResourceItem = {
```
Comment (ushkal) [inline packages/dx-scheduler-core/src/types/resources.types.ts]: Plane -> Plain ?

Comment (MaximKudriavtsev) [reply packages/dx-scheduler-core/src/types/resources.types.ts]: Sure

File: packages/dx-react-scheduler-demos/src/demo-sources/scheduler-resources/material-ui/editing.jsx [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+            defaultCurrentDate="2017-05-25"
+          />
+          <EditingState
+            onCommitChanges={this.commitChanges}
+          />
+          <EditRecurrenceMenu />
+
+          <MonthView />
+          <Appointments />
+          <AppointmentTooltip />
```
Comment (AryamnovEugeniy) [inline packages/dx-react-scheduler-demos/src/demo-sources/scheduler-resources/material-ui/editing.jsx]: I guess, we should add Open Button

File: packages/dx-react-scheduler-material-ui/src/templates/appointment-form/basic/layout.jsx [hunk 3]
```diff
@@ -69,6 +75,14 @@ const styles = ({ spacing, typography }) => ({
...
     marginTop: spacing(0.875),
   },
+  chips: {
+    display: 'flex',
+    flexWrap: 'wrap',
+  },
+  chip: {
+    color: 'white',
+    margin: 2,
+  },
```
Comment (AryamnovEugeniy) [inline packages/dx-react-scheduler-material-ui/src/templates/appointment-form/basic/layout.jsx]: Unused styles

File: packages/dx-react-scheduler-demos/src/demo-data/tasks.js [hunk 4]
```diff
@@ -240,15 +240,15 @@ export const owners = [
...
   }, {
     text: 'Taylor Riley',
     id: 4,
-    color: '#E91E63',
+    color: '#006064',
   }, {
     text: 'Brad Farkus',
     id: 5,
-    color: '#AB47BC',
+    color: '#004D40',
```
Comment (AryamnovEugeniy) [inline packages/dx-react-scheduler-demos/src/demo-data/tasks.js]: Should we change the Overview Demo as well?

File: packages/dx-react-scheduler/src/plugins/drag-drop-provider.tsx [hunk 5]
```diff
@@ -266,17 +268,19 @@ class DragDropProviderBase extends React.PureComponent<
...
-            >
-              {payload && params.data.id === payload.id ? (
-                <SourceAppointment {...params} />
-              ) : (
-                <TemplatePlaceholder params={{ ...params, draggable: true }} />
-              )}
-            </DragSource>
-          )}
+          {({ styles, ...params }: any) => {
+            return (
```
Comment (AryamnovEugeniy) [inline packages/dx-react-scheduler/src/plugins/drag-drop-provider.tsx]: We shouldn't use return expression here

File: packages/dx-scheduler-core/src/plugins/resources/computeds.test.ts [hunk 6]
```diff
@@ -0,0 +1,209 @@
...
+              allowMultiple: false,
+              isMain: true,
+              text: 'ownerId',
+              title: 'ownerId',
+              fieldName: 'ownerId',
+            }],
+          },
+        ]);
+    });
+    it('should assign the Main flag to two resource', () => {
```
Comment (AryamnovEugeniy) [inline packages/dx-scheduler-core/src/plugins/resources/computeds.test.ts]: 'should not assign the Main flag to two resources'?
Only the first, `ownerId`, resource has the `isMain` prop equal to `true`

**Comment by MaximKudriavtsev** [comment]
FPR

File: packages/dx-react-scheduler-material-ui/src/templates/appointment/appointment.jsx [hunk 7]
```diff
@@ -63,19 +71,20 @@ const AppointmentBase = ({
...
 AppointmentBase.defaultProps = {
+  resources: [],
   onClick: undefined,
   className: undefined,
   data: {},
   draggable: false,
 };
 
-export const Appointment = withStyles(styles, { name: 'Appointment' })(AppointmentBase);
+export const Appointment = AppointmentBase;
```
Comment (AryamnovEugeniy) [inline packages/dx-react-scheduler-material-ui/src/templates/appointment/appointment.jsx]: We do not need it. It's better to rename AppointmentBase on 35th line

File: packages/dx-react-scheduler-material-ui/src/templates/appointment/appointment.test.jsx [hunk 8]
```diff
@@ -65,7 +67,7 @@ describe('Appointment', () => {
         >
           <div />
         </Appointment>
-      )).find(`.${classes.appointment}`);
+      )).find('.appointment');
```
Comment (AryamnovEugeniy) [inline packages/dx-react-scheduler-material-ui/src/templates/appointment/appointment.test.jsx]: Why do we have to look for '.appointment'? The root component is '.appointment'

File: packages/dx-react-scheduler/docs/reference/resources.md [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+color? | string &#124; [Color](#color) | The resource item color that would used inside UI components.
+
+### Palette
+
+Specifies the resources palette.
+
+Type: `Array<string | Color>`
+
+### ValidResource
+
```
Comment (AryamnovEugeniy) [inline packages/dx-react-scheduler/docs/reference/resources.md]: Should we add a description here?

File: packages/dx-react-scheduler/docs/reference/resources.md [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+fieldName | string | The appointment data field name that would binding to resource.
+text | string | The resource item text that would displays inside UI components.
+title | string | The resource kind title that would displays inside UI components.
+allowMultiple | boolean | Indicates whether or not several resources of this kind can be assigned to an appointment.
+isMain | boolean | Specifies the main resource kind.
+items | Array&lt;[ValidResourceItem](#validresourceitem)&gt; | The array of resource items.
+color? | string &#124; [Color](#color) | The resource item color that would used inside UI components.
+
+### ValidResourceItem
+
```
Comment (AryamnovEugeniy) [inline packages/dx-react-scheduler/docs/reference/resources.md]: Same here

File: packages/dx-react-scheduler-material-ui/src/templates/appointment-form/basic/layout.jsx [hunk 11]
```diff
@@ -198,6 +199,22 @@ const LayoutBase = ({
         onValueChange={changeNotes}
         className={classes.notesEditor}
       />
+      {resources.map(resource => (
+        <>
```
Comment (AryamnovEugeniy) [inline packages/dx-react-scheduler-material-ui/src/templates/appointment-form/basic/layout.jsx]: Seems that we need to add `key` here
image

File: packages/dx-react-scheduler/docs/reference/resources.md [hunk 12]
```diff
@@ -0,0 +1,91 @@
...
+### Dependencies
+
+- [Appointments](appointments.md)
+
+### Properties
+
+Name | Type | Default | Description
+-----|------|---------|------------
+data | Array&lt;[Resource](#resource)&gt; | The resources data set.
```
Comment (AryamnovEugeniy) [inline packages/dx-react-scheduler/docs/reference/resources.md]: Description is in Default's place

File: packages/dx-react-scheduler-material-ui/src/templates/appointment-tooltip/content.jsx [hunk 13]
```diff
@@ -65,17 +73,18 @@ const styles = ({ spacing, palette, typography }) => ({
     width: '100%',
     height: '100%',
   },
-});
+}));
 
 export const ContentBase = ({
```
Comment (LazyLahtak) [inline packages/dx-react-scheduler-material-ui/src/templates/appointment-tooltip/content.jsx]: `export const ContentBase` -> `export const Content`

File: packages/dx-react-scheduler-demos/src/demo-sources/scheduler-resources/material-ui/basic.jsx [hunk 14]
```diff
@@ -0,0 +1,159 @@
...
+            { id: 5, text: 'Brad Farkus' },
+          ],
+        },
+      ],
+    };
+
+    this.changeMainResource = this.changeMainResource.bind(this);
+  }
+
+  changeMainResource(mainResourceName) {
```
Comment (ushkal) [inline packages/dx-react-scheduler-demos/src/demo-sources/scheduler-resources/material-ui/basic.jsx]: This function can be declared in constructor using fat arrow then binding is unnecessary

File: packages/dx-react-scheduler-material-ui/src/templates/appointment-form/basic/layout.jsx [hunk 15]
```diff
@@ -198,6 +199,22 @@ const LayoutBase = ({
         onValueChange={changeNotes}
         className={classes.notesEditor}
       />
+      {resources.map(resource => (
```
Comment (ushkal) [inline packages/dx-react-scheduler-material-ui/src/templates/appointment-form/basic/layout.jsx]: We should test it too

File: packages/dx-react-scheduler-material-ui/src/templates/utils.js [hunk 16]
```diff
@@ -30,3 +30,15 @@ export const scrollingStrategy = (scrollablePart, fixedPart) => {
...
+export const getResourceColor = (resources) => {
+  if (resources && resources.length) {
+    return resources.find(resource => resource.isMain).color;
+  } return undefined;
+};
+
+export const getAppointmentColor = (level, color, defaultColor) => {
+  if (!color) return setColor(level, defaultColor);
+  if (typeof color === 'string') return color;
+  return setColor(level, color);
```
Comment (ushkal) [inline packages/dx-react-scheduler-material-ui/src/templates/utils.js]: `setColor` does not actually sets color. May be better `ensureColor` or similar?

File: packages/dx-scheduler-core/src/types/resources.types.ts [hunk 17]
```diff
@@ -0,0 +1,90 @@
...
+   * This id should be related with scheduler's data `fieldName` id.
+   */
+  id: number | string;
+  /**
+   * Specifies resource item's color.
+   * It is used to indicate appointments related to this resource.
+   */
+  color: string | Color;
+  /** Specifies the resource item's text that would be used in UI */
+  text: string;
```
Comment (ushkal) [inline packages/dx-scheduler-core/src/types/resources.types.ts]: `ValidResourceItem` can inherit the `ResourceItem`: 
```
ValidResourceItem = Required<ResourceItem> & { title, firldName, ... }
```

File: packages/dx-scheduler-core/src/types/resources.types.ts [hunk 18]
```diff
@@ -0,0 +1,90 @@
...
+export type ValidResourceItem = {
+  /**
+   * Specifies a resource item's identifier.
+   * This id should be related with scheduler's data `fieldName` id.
+   */
+  id: number | string;
+  /**
+   * Specifies resource item's color.
+   * It is used to indicate appointments related to this resource.
+   */
```
Comment (ushkal) [inline packages/dx-scheduler-core/src/types/resources.types.ts]: I think it's not good to reference theme in a core package. Do we need mui color?

Comment (MaximKudriavtsev) [reply packages/dx-scheduler-core/src/types/resources.types.ts]: I think yeah, we need it. Integration with Material-UI is our main (only one) advantage.
This scheduler has only one theme. I think we can close our eyes and do it.

Comment (ushkal) [reply packages/dx-scheduler-core/src/types/resources.types.ts]: This comment addresses this line:
```export interface Color extends MUIColor {}```

Comment (ushkal) [reply packages/dx-scheduler-core/src/types/resources.types.ts]: Ok, let it be so :)

**Comment by MaximKudriavtsev** [comment]
FPR

File: packages/dx-react-scheduler/docs/reference/resources.md [hunk 19]
```diff
@@ -0,0 +1,93 @@
...
+
+- [Appointments](appointments.md)
+
+### Properties
+
+Name | Type | Default | Description
+-----|------|---------|------------
+data | Array&lt;[Resource](#resource)&gt; | | The resources data set.
+mainResourceName? | string | | Specifies whether appointments are colored like this resource kind.
+palette? | [Palette](#palette) | [red, pink, purple, deepPurple, indigo, blue, lightBlue, cyan, teal, green, lightGreen, lime, yellow, ember, orange, deepOrange] | The palette that would be used if resource item colors were not defined.
```
[the remark this answers is not in the document]
Comment (MaximKudriavtsev) [reply packages/dx-react-scheduler/docs/reference/resources.md]: You can set a color by a `string` variable.

[the remark this answers is not in the document]
Comment (MaximKudriavtsev) [reply packages/dx-react-scheduler/docs/reference/resources.md]: No, these list contains Material-UI color objects

File: packages/dx-react-scheduler/docs/guides/resources.md [hunk 20]
```diff
@@ -0,0 +1,33 @@
...
+
+To assign appointments to resources, users can edit a dedicated [AppointmentForm](../reference/appointment-form.md) field.
+
+## Related Plugins
+
+- [Resources](../reference/resources.md) - configures resources
+
+## Single and Multiple Instance Resources
+
+Appointments can be assigned to multiple or only one instance of a resource, depending on the resource type. To illustrate the difference, let us return to the example in which appointments are talks at a conference and resources are rooms, speakers, and attendees.
```
Comment (dirk-pieterse) [inline packages/dx-react-scheduler/docs/guides/resources.md]:
```suggestion
Appointments can be assigned to single/multiple resource instances, depending on the resource type. The same example about a conference can be used to illustrate the difference:
```

File: packages/dx-react-scheduler/docs/guides/resources.md [hunk 21]
```diff
@@ -0,0 +1,33 @@
...
+
+A talk can only be held in a single room, so the appointment can only be assigned to one instance of the Rooms resource. Such resources are called **single instance resources**. This is the default resource type.
```
Comment (dirk-pieterse) [inline packages/dx-react-scheduler/docs/guides/resources.md]:
```suggestion
A talk can only be held in one room, and the appointment can only be assigned to one instance of the Rooms resource. These are default resource types called **single instance resources**.
```

File: packages/dx-react-scheduler/docs/reference/appointment-form.md [hunk 22]
```diff
@@ -110,6 +113,7 @@ dateEditorComponent | ComponentType&lt;[AppointmentForm.DateEditorProps](#appoin
 booleanEditorComponent | ComponentType&lt;[AppointmentForm.BooleanEditorProps](#appointmentformbooleanteditorprops)&gt; | A component that renders an editor of Boolean values.
 selectComponent | ComponentType&lt;[AppointmentForm.SelectProps](#appointmentformselectprops)&gt; | A component that renders a menu of options.
```
Comment (dirk-pieterse) [inline packages/dx-react-scheduler/docs/reference/appointment-form.md]:
```suggestion
selectComponent | ComponentType&lt;[AppointmentForm.SelectProps](#appointmentformselectprops)&gt; | A component that renders an options menu.
```

File: packages/dx-react-scheduler/docs/reference/appointment-form.md [hunk 23]
```diff
@@ -101,6 +102,8 @@ Field | Type | Description
 ------|------|------------
 readOnly? | boolean | Specifies whether the appointment form is read-only.
 appointmentData | [AppointmentModel](./scheduler.md#appointmentmodel) | The appointment's data.
+appointmentResources | Array&lt;[ValidResourceInstance](./resources.md#validresourceinstance)&gt; | The appointment's resource instances.
+resources | Array&lt;[ValidResource](./resources.md#validresource)&gt; | All resources.
 onFieldChange | (nextFieldValue: { [fieldName: string]: any }) => void | An event raised when a field value in the appointment form is changed.
```
Comment (dirk-pieterse) [inline packages/dx-react-scheduler/docs/reference/appointment-form.md]:
```suggestion
onFieldChange | (nextFieldValue: { [fieldName: string]: any }) => void | An event this is raised when a field value in the appointment form is changed.
```

File: packages/dx-react-scheduler/docs/reference/resources.md [hunk 24]
```diff
@@ -0,0 +1,93 @@
...
+
+Field | Type | Description
+------|------|------------
+id | number &#124; string | The resource instance ID.
+text? | string | The resource instance text.
+color? | string &#124; [Color](#color) | The resource instance color.
+
+### Palette
+
+Specifies a palette that provides colors for those resource instances whose color is not defined.
```
Comment (dirk-pieterse) [inline packages/dx-react-scheduler/docs/reference/resources.md]:
```suggestion
Specifies a palette that provides colors for resource instances with undefined colors.
```

**Pull request merged**